File: src/main/java/com/tngtech/jgiven/reference/JGivenUsageTypeProvider.java
package com.tngtech.jgiven.reference;

import com.intellij.psi.PsiElement;
import com.intellij.usages.impl.rules.UsageType;
import com.intellij.usages.impl.rules.UsageTypeProvider;
import com.tngtech.jgiven.scenario.state.ScenarioStateAnnotationProvider;
import org.jetbrains.annotations.Nullable;

public class JGivenUsageTypeProvider implements UsageTypeProvider {
    static final String USAGE_TYPE = "JGiven Scenario State";
    private ScenarioStateAnnotationProvider scenarioStateAnnotationProvider = new ScenarioStateAnnotationProvider();

    @Nullable
    @Override
    public UsageType getUsageType(PsiElement element) {
        if (scenarioStateAnnotationProvider.isJGivenScenarioState(element)) {
            return new UsageType(USAGE_TYPE);
        }
        return null;
    }
}

File: src/main/java/com/tngtech/jgiven/reference/JGivenUsageTypeProvider.kt
package com.tngtech.jgiven.reference

import com.intellij.psi.PsiElement
import com.intellij.usages.impl.rules.UsageType
import com.intellij.usages.impl.rules.UsageTypeProvider
import com.tngtech.jgiven.scenario.state.ScenarioStateAnnotationProvider

class JGivenUsageTypeProvider(
        private val scenarioStateAnnotationProvider: ScenarioStateAnnotationProvider = ScenarioStateAnnotationProvider()
) : UsageTypeProvider {
    val usageType = UsageType {"JGiven Scenario State"}
    override fun getUsageType(element: PsiElement) =
            when {
                scenarioStateAnnotationProvider.isJGivenScenarioState(element) -> usageType
                else -> null
            }
}
